internal/daemon/controller/common/scopeids/scope_ids.go

```diff
@@ -353,3 +353,143 @@ func GetListingScopeIds(
 	}
 	return scopeResourceInfo.ScopeIds, scopeMap, nil
 }
+
+// AuthorizedForList retrieves and returns all scopes where a user is authorized
+// to perform a *list* action on. It looks recursively from `rootScopeId`.
+func AuthorizedForList(
+	ctx context.Context, // The context to use when listing in the DB, if required
+	iamRepoFn common.IamRepoFactory, // An IAM repo function to use for a scope listing call
+	authResults auth.VerifyResults, // The original auth results from the list command
+	rootScopeId string, // The scope ID to use, or the starting point for a recursive search
+	resourceType resource.Type, // The type of resource being listed
+	recursive bool, // Whether the search is recursive.
+) (map[string]*scopes.ScopeInfo, error) {
+	const op = "scopeids.AuthorizedForList"
+
+	// Validation
+	switch {
+	case resourceType == resource.Unknown:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "unknown resource")
+	case iamRepoFn == nil:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "nil iam repo")
+	case rootScopeId == "":
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing root scope id")
+	case authResults.Scope == nil:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "nil scope in auth results")
+	case !recursive && authResults.Scope.Id != rootScopeId:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "non-recursive search but auth results scope does not match root scope")
+	}
+
+	// This will be used to save scope info so we can put the right scope
+	// info for each returned value
+	scopeResourceMap := make(map[string]*scopes.ScopeInfo)
+
+	// Base case: if not recursive, return the scope we were given and the
+	// already-looked-up info.
+	if !recursive {
+		return map[string]*scopes.ScopeInfo{authResults.Scope.Id: authResults.Scope}, nil
+	}
```

Comment:
likewise this can be dropped if we drop the `recursive` arg

Reply:
https://github.com/hashicorp/boundary/pull/2336#discussion_r941161611